pkg/uefi/flash.go

```diff
@@ -213,59 +196,77 @@ func NewFlashImage(buf []byte) (*FlashImage, error) {
 		return nil, err
 	}
 
-	// Add to extractable regions
-	f.regions = append(f.regions, &f.IFD)
+	// FlashRegions
+	frs := f.IFD.Region.FlashRegions
 
-	// BIOS region
-	if !f.IFD.Region.BIOS.Valid() {
-		return nil, fmt.Errorf("no BIOS region: invalid region parameters %v", f.IFD.Region.BIOS)
-	}
-	br, err := NewBIOSRegion(buf[f.IFD.Region.BIOS.BaseOffset():f.IFD.Region.BIOS.EndOffset()], &f.IFD.Region.BIOS)
-	if err != nil {
-		return nil, err
-	}
-	f.BIOS = br
-	// Add to extractable regions
-	f.regions = append(f.regions, f.BIOS)
-
-	// ME region
-	if f.IFD.Region.ME.Valid() {
-		mer, err := NewMERegion(buf[f.IFD.Region.ME.BaseOffset():f.IFD.Region.ME.EndOffset()], &f.IFD.Region.ME)
-		if err != nil {
-			return nil, err
-		}
-		f.ME = mer
-		// Add to extractable regions
-		f.regions = append(f.regions, f.ME)
+	// BIOS region has to be valid
+	if !frs[BIOS].Valid() {
+		return nil, fmt.Errorf("no BIOS region: invalid region parameters %v", frs[BIOS])
 	}
 
-	// GBE region
-	if f.IFD.Region.GBE.Valid() {
-		gber, err := NewGBERegion(buf[f.IFD.Region.GBE.BaseOffset():f.IFD.Region.GBE.EndOffset()], &f.IFD.Region.GBE)
-		if err != nil {
-			return nil, err
+	// Parse all the regions
+	for i, fr := range frs {
+		if !fr.Valid() {
+			continue
+		}
+		if o := uint64(fr.BaseOffset()); o > f.FlashSize {
+			return nil, fmt.Errorf("region out of bounds: BaseOffset %#x, Flash size %#x",
+				o, f.FlashSize)
+		}
+		if o := uint64(fr.EndOffset()); o > f.FlashSize {
+			return nil, fmt.Errorf("region out of bounds: EndOffset %#x, Flash size %#x",
+				o, f.FlashSize)
+		}
+		if c, ok := regionConstructors[FlashRegionType(i)]; ok {
+			r, err := c(buf[fr.BaseOffset():fr.EndOffset()], &frs[i], FlashRegionType(i))
+			if err != nil {
+				return nil, err
+			}
+			f.Regions = append(f.Regions, r)
 		}
-		f.GBE = gber
-		// Add to extractable regions
-		f.regions = append(f.regions, f.GBE)
 	}
 
-	// PD region
-	if f.IFD.Region.PD.Valid() {
-		pdr, err := NewPDRegion(buf[f.IFD.Region.PD.BaseOffset():f.IFD.Region.PD.EndOffset()], &f.IFD.Region.PD)
-		if err != nil {
-			return nil, err
+	// Sort the regions by offset so we can look for gaps
+	sort.Slice(f.Regions, func(i, j int) bool {
+		return f.Regions[i].FlashRegion().Base < f.Regions[j].FlashRegion().Base
+	})
+
+	// Search for gaps and fill in with unknown regions
+	offset := uint64(FlashDescriptorLength)
+	var newRegions []Region
+	for _, r := range f.Regions {
+		nextBase := uint64(r.FlashRegion().BaseOffset())
+		if nextBase < offset {
+			// Something is wrong, overlapping regions
+			// TODO: print a better error message describing what it overlaps with
+			return nil, fmt.Errorf("overlapping regions! region %v overlaps with the previous region", r)
 		}
-		f.PD = pdr
-		// Add to extractable regions
-		f.regions = append(f.regions, f.PD)
+		if nextBase > offset {
+			// There is a gap, create an unknown region
+			tempFR := &FlashRegion{Base: uint16(offset / RegionBlockSize),
```

Comment:
Would it be worth writing a unit test for this?

Reply:
I'll have to create a sample image with some fake ifd. Might be too much for this PR